rfcs/026-work-enricher/README.md

```diff
@@ -0,0 +1,160 @@
+
+```
+A
+├── 1
+│   ├── X
+│   └── Y
+│
+└── 2
+```
+
+Generally, for the insertion, update or removal of any given node, we need to update its siblings, its parent, and the whole subtree containing its descendants. The worst case for this process is an update of the root node, in which case the whole archive needs to be denormalised, with the largest archives containing in the order of 1000 works.
+
+## Commonalities with Other Pipeline Stages
+
+The matcher and merger stages have similar underlying functionality to the enricher. Abstractly these stages involve:
+
+1) Receiving a work as input and applying domain knowledge to build a graph of connections to other works
+2) Using the input work and its graph to emit a new work at the output, including any required data from the relations
+
+This points to a potential universal solution within the pipeline, such as using a common service, library or data store to assist with some of these tasks. Additionally, there are potential similarities with upcoming concepts work, where there has been some investigation into how concepts would fit within a graph model.
+
+There are a few issues to bear in mind when considering how a shared solution would work:
+
+* Whilst stage 1 (currently the matcher) could potentially also involve analysis of archive relations, it is important as stated above that the denormalisation in the enricher takes place on a merged works to ensure data consistency. This indicates the merger and enricher stages being separate and accessing works from two separate works stores (unmerged and merged).
+
+* The matcher currently uses a custom DynamoDB graph implementation which was built specifically for the purpose. This was never meant to be a general solution and a rewrite would likely need to be involved in the building of some common system.
+
+* The graph element of both the matcher and relations work are relatively similar, but the denormalisation stages (the enricher and the merger) have less parallels. For example, the input to the merger is currently a group of work IDs to be merged with the stage having no knowledge of the graph representation, unlike the enricher which is likely to receive single works and needs access to the graph.
+
+Besides from the use of graphs and similarities to the matcher and merger, another aspect worth mentioning is the fact that we will need another store for denormalised works in addition to a store for merged works. Having two Elasticsearch indices coupled to corresponding ingestors is a possible solution (one for the current index of merged works and adding an additional one for denormalised works). There have been some previous discussions about replacing the recorder VHS with an Elasticsearch index, so it might be a worthwhile time to consider how this would work in this context, as here we would introducing an Elasticsearch index which was not the sink of the pipeline but used as an intermediate store.
```

Comment:
Not saying that we should address this as part of this RFC but this might also sit within the more general problem of "how do we store changes to works in the pipeline" which we currently solve in one way in the recorder and in another way in big messaging

Reply:
Yes completely, I think it would be good to get it right for this work so up for having these discussions before proceeding if people think it worthwhile.

Reply:
i think this, and how it affects denormalising changes between works/concepts, is something that it would be especially useful to get @jamesgorrie's opinion on. we went back and forth on a lot of options while fleshing out the concepts branch, but a solution which works for the whole pipeline would be amazing

Reply:
IMO we should switch out the VHS for a faster cheaper store in the recorders. The question is when. Do we do it now to make future reindexes faster? Or do we wait until we've met the library switch-off objective? Will making the change now enable us to reach that objective faster (given the dev time to switch the recorder datastore, risk of dealing with smoothing out bugs with the "new" way).

Reply:
> Will making the change now enable us to reach that objective faster?

Depends on how many reindexes we expect to do between now and the library switch-off, right? If the index is fairly stable and we're not reindexing very often, it can wait. If we expect to make a lot of changes between now and the switch off, doing this work now might save us time later.

Reply:
I agree we should switch out the recorder VHS only when we have proved the concepts elsewhere and there is more time available for optimising the pipeline. But it makes sense to come up with a solution for this work which we think will be applicable to that too IMO.

Reply:
Just a thought experiment here, what if we have 1 works index that has multiple services writing to it?

Some concerns:
* debugging
* re-running from different steps in the pipeline
* ?...

![1indextorulethemall(1)](https://user-images.githubusercontent.com/31692/89423490-79ca7400-d72e-11ea-85bc-c1d0b6f8431a.png)

Reply:
The more I look at this ☝️ and race conditions and a whole bunch of other things, it feels like a bad idea.

Reply:
Why would we want to update just one index though? I don't think I understand what problem would we solve this way... Or maybe I don't understand the proposal

Reply:
What I suggested was definitely a bad idea, thinking aloud.